File: scripts/find-onsets.mjs
#!/usr/bin/env node
/**
 * find-onsets.mjs
 *
 * Detects track boundaries in a downloaded MP3 using ffmpeg astats +
 * first-derivative onset detection. Overwrites startTime on all tracks
 * in episodes.json with verified-candidate timestamps.
 *
 * How it works:
 *   1. Run ffmpeg astats (50ms windows) on the full MP3 → per-frame RMS in dB
 *   2. Smooth with a sliding median (~500ms window) to reduce transient noise
 *   3. Compute first derivative (frame-to-frame delta of smoothed RMS)
 *   4. Find local maxima above DERIVATIVE_THRESHOLD, enforcing MIN_TRACK_GAP_SECS
 *   5. Select exactly (trackCount - 1) boundaries by peak prominence
 *   6. Score each onset by strength + isolation → confidence 0–1
 *   7. Write results to episodes.json (unless --dry-run)
 *
 * Usage:
 *   node scripts/find-onsets.mjs 1975
 *   node scripts/find-onsets.mjs 1975 --dry-run        # print only, no write
 *   node scripts/find-onsets.mjs 1975 --all-candidates # show all peaks before selection
 */

import { readFileSync, writeFileSync, existsSync } from 'fs';
import { resolve, dirname } from 'path';
import { fileURLToPath } from 'url';
import { spawnSync } from 'child_process';

const __dirname = dirname(fileURLToPath(import.meta.url));
const ROOT      = resolve(__dirname, '..');
const DATA_FILE = resolve(ROOT, 'public/data/episodes.json');
const AUDIO_DIR = resolve(ROOT, 'public/audio');

// ─── Tuning constants ──────────────────────────────────────────────────────────
const FRAME_WINDOW_SECS    = 0.05;  // astats integration window (50ms) — controls smoothing
                                     // NOTE: output frame rate is the codec's native rate
                                     // (~23ms for 44.1kHz/1024-sample frames), not this value.
                                     // minGapFrames and smooth window are derived from actual data.
const SMOOTH_WINDOW_SECS   = 0.5;   // target smoothing window in seconds (~500ms)
const DERIVATIVE_THRESHOLD = 2.0;   // dB/frame minimum delta for onset candidate
const MIN_TRACK_GAP_SECS   = 10;    // minimum seconds between detected tracks
// ──────────────────────────────────────────────────────────────────────────────

// ─── Parse CLI args ────────────────────────────────────────────────────────────
const args          = process.argv.slice(2);
const episodeArg    = args.find(a => /^\d+$/.test(a));
const dryRun        = args.includes('--dry-run');
const allCandidates = args.includes('--all-candidates');

if (!episodeArg) {
  console.error('Usage: node scripts/find-onsets.mjs <episodeId> [--dry-run] [--all-candidates]');
  process.exit(1);
}
const episodeId = parseInt(episodeArg, 10);
// ──────────────────────────────────────────────────────────────────────────────

// Load episodes.json
let data;
try {
  data = JSON.parse(readFileSync(DATA_FILE, 'utf-8'));
} catch (err) {
  console.error(`Error reading episodes.json: ${err.message}`);
  process.exit(1);
}

const episode = data.episodes.find(ep => ep.id === episodeId);
if (!episode)           { console.error(`Episode ${episodeId} not found.`);       process.exit(1); }
if (!episode.tracks?.length) { console.error(`Episode ${episodeId} has no tracks.`); process.exit(1); }

const mp3File = resolve(AUDIO_DIR, `mrr-radio-${episodeId}.mp3`);
if (!existsSync(mp3File)) {
  console.error(`Error: ${mp3File} not found.`);
  console.error(`Run: node scripts/download-mp3.mjs ${episodeId}`);
  process.exit(1);
}

const trackCount      = episode.tracks.length;
const boundariesNeeded = trackCount - 1; // track 0 always starts at 0

console.log(`\nEpisode #${episodeId}: ${trackCount} tracks, ${fmtTime(episode.durationSecs)} total`);
console.log(`Need to detect ${boundariesNeeded} track boundaries.`);
console.log(`Running ffmpeg astats (frame=${FRAME_WINDOW_SECS}s)...\n`);

// ─── Main pipeline ─────────────────────────────────────────────────────────────
const frames = extractFrameRMS(mp3File);

// Derive actual frame duration from pts_time values (codec native rate, not FRAME_WINDOW_SECS)
const actualFrameDuration = frames.length > 1
  ? (frames[frames.length - 1].time - frames[0].time) / (frames.length - 1)
  : FRAME_WINDOW_SECS;
const smoothWindowFrames = Math.max(3, Math.round(SMOOTH_WINDOW_SECS / actualFrameDuration));
const minGapFrames       = Math.ceil(MIN_TRACK_GAP_SECS / actualFrameDuration);

console.log(`Extracted ${frames.length} frames (actual frame: ${(actualFrameDuration * 1000).toFixed(1)}ms, smooth: ${smoothWindowFrames} frames, gap: ${minGapFrames} frames)`);

const rmsValues  = frames.map(f => f.rms);
const smoothed   = smoothMedian(rmsValues, smoothWindowFrames);
const derivative = firstDerivative(smoothed);

const candidates = findOnsetCandidates(frames, derivative, minGapFrames, DERIVATIVE_THRESHOLD);
console.log(`Found ${candidates.length} onset candidates (threshold=${DERIVATIVE_THRESHOLD}, gap=${MIN_TRACK_GAP_SECS}s)`);

if (allCandidates) {
  console.log('\nAll candidates:');
  candidates.forEach((c, i) => {
    console.log(`  ${String(i + 1).padStart(3)}. ${fmtTime(c.time).padEnd(10)} delta=${c.delta.toFixed(2).padStart(6)} dB/frame`);
  });
  console.log('');
}

let selected;
if (candidates.length >= boundariesNeeded) {
  selected = selectTopN(candidates, boundariesNeeded);
  console.log(`Selected top ${boundariesNeeded} by peak prominence, sorted chronologically.`);
} else {
  selected = [...candidates].sort((a, b) => a.time - b.time);
  console.log(`\n⚠  Only ${candidates.length} candidates found (need ${boundariesNeeded}).`);
  console.log(`   Lower DERIVATIVE_THRESHOLD (currently ${DERIVATIVE_THRESHOLD}) and re-run.`);
}

const scored = scoreOnsets(selected, episode.durationSecs);
const sorted  = [...episode.tracks].sort((a, b) => a.trackIndex - b.trackIndex);
assignTimestamps(sorted, scored, episode.durationSecs);

printTable(sorted, episode.durationSecs);

if (!dryRun) {
  episode.tracks = episode.tracks.map(t => sorted.find(s => s.id === t.id) || t);
  writeFileSync(DATA_FILE, JSON.stringify(data, null, 2));
  console.log('\nWritten to episodes.json.');
  console.log('Run again with --dry-run to preview without writing.');
} else {
  console.log('\n--dry-run mode: episodes.json was NOT modified.');
}
// ──────────────────────────────────────────────────────────────────────────────

/**
 * Run ffmpeg astats on the MP3 and return an array of { time, rms } objects,
 * one per 50ms frame. rms is in dB (-100 for silence, -inf coerced to -100).
 *
 * We filter ametadata output to a single key (Overall.RMS_level) to keep
 * stdout small (~7 MB for a 57-min episode vs ~220 MB for all keys).
 * Frame times are derived from pts_time when present, or computed as
 * frameIndex × FRAME_WINDOW_SECS as a fallback.
 */
function extractFrameRMS(mp3File) {
  const rmsKey = 'lavfi.astats.Overall.RMS_level';
  const result = spawnSync('ffmpeg', [
    '-i', mp3File,
    '-af', `astats=length=${FRAME_WINDOW_SECS}:metadata=1:reset=1,ametadata=mode=print:key=${rmsKey}:file=-`,
    '-f', 'null', '-',
  ], { encoding: 'utf-8', maxBuffer: 50 * 1024 * 1024 });

  if (result.error) {
    console.error(`ffmpeg error: ${result.error.message}`);
    process.exit(1);
  }
  if (result.status !== 0 && !result.stdout) {
    console.error('ffmpeg failed with no output. Is ffmpeg installed?');
    process.exit(1);
  }

  const frames = [];
  let currentTime = null;
  let frameIndex  = 0;

  for (const line of (result.stdout || '').split('\n')) {
    // Frame descriptor line: "frame:N  pts:P  pts_time:T"
    const timeMatch = line.match(/pts_time:([\d.]+)/);
    if (timeMatch) {
      currentTime = parseFloat(timeMatch[1]);
      continue;
    }
    // RMS line: "lavfi.astats.Overall.RMS_level=VALUE"
    if (line.startsWith(rmsKey + '=')) {
      const raw = line.slice(rmsKey.length + 1).trim();
      const rms = (raw === '-inf' || raw === 'inf') ? -100 : parseFloat(raw);
      // Use pts_time if available, else derive from frame count
      const time = currentTime !== null ? currentTime : frameIndex * FRAME_WINDOW_SECS;
      frames.push({ time, rms });
      currentTime = null;
      frameIndex++;
    }
  }

  if (frames.length === 0) {
    console.error('No frames extracted. Check ffmpeg output and MP3 path.');
    process.exit(1);
  }

  return frames;
}

/**
 * Sliding median smoothing. Reduces transient noise in the RMS curve.
 * windowSize frames on each side (centered window of ~windowSize+1 frames).
 */
function smoothMedian(values, windowSize) {
  const half = Math.floor(windowSize / 2);
  return values.map((_, i) => {
    const start  = Math.max(0, i - half);
    const end    = Math.min(values.length, i + half + 1);
    const window = values.slice(start, end).slice().sort((a, b) => a - b);
    return window[Math.floor(window.length / 2)];
  });
}

/**
 * First derivative: frame-to-frame delta of the smoothed RMS curve.
 * A large positive value indicates a sudden energy increase (track attack).
 */
function firstDerivative(values) {
  return values.map((v, i) => i === 0 ? 0 : v - values[i - 1]);
}

/**
 * Find all local maxima of the derivative that exceed the threshold,
 * then enforce minimum gap by keeping only the strongest peak in each window.
 * Returns candidates sorted by time.
 */
function findOnsetCandidates(frames, derivative, minGapFrames, threshold) {
  // Find local maxima above threshold
  const rawPeaks = [];
  for (let i = 1; i < derivative.length - 1; i++) {
    if (
      derivative[i] > threshold &&
      derivative[i] >= derivative[i - 1] &&
      derivative[i] >= derivative[i + 1]
    ) {
      rawPeaks.push({ frameIdx: i, time: frames[i].time, delta: derivative[i] });
    }
  }

  // Greedy minimum-gap enforcement: sort by strength, keep strongest non-conflicting peaks
  rawPeaks.sort((a, b) => b.delta - a.delta);
  const kept = [];
  for (const peak of rawPeaks) {
    const tooClose = kept.some(k => Math.abs(k.frameIdx - peak.frameIdx) < minGapFrames);
    if (!tooClose) kept.push(peak);
  }

  kept.sort((a, b) => a.time - b.time);
  return kept;
}

/**
 * Select top N candidates by peak delta, then sort chronologically.
 */
function selectTopN(candidates, n) {
  return [...candidates]
    .sort((a, b) => b.delta - a.delta)
    .slice(0, n)
    .sort((a, b) => a.time - b.time);
}

/**
 * Score each detected onset with a confidence value 0–1.
 * Based on: peak strength (70%) + isolation from neighbors (30%).
 */
function scoreOnsets(onsets, episodeDurationSecs) {
  if (onsets.length === 0) return [];

  const maxDelta = Math.max(...onsets.map(o => o.delta));

  return onsets.map((onset, i) => {
    const strengthScore = onset.delta / maxDelta;

    const prevTime = i === 0 ? 0 : onsets[i - 1].time;
    const nextTime = i === onsets.length - 1 ? episodeDurationSecs : onsets[i + 1].time;
    const minGap   = Math.min(onset.time - prevTime, nextTime - onset.time);
    // 60s gap → isolation score of 1.0; shorter gaps → proportionally lower
    const isolationScore = Math.min(1, minGap / 60);

    const confidence = Math.round((0.7 * strengthScore + 0.3 * isolationScore) * 100) / 100;
    return { ...onset, confidence };
  });
}

/**
 * Write startTime, startTimeMethod, startTimeConfidence, and durationSecs
 * onto each track object. Track 0 always gets startTime=0, method="manual".
 * tracks must be sorted by trackIndex.
 */
function assignTimestamps(tracks, scoredOnsets, episodeDurationSecs) {
  // onset i (0-indexed) → track index i+1
  const onsetMap = {};
  scoredOnsets.forEach((o, i) => { onsetMap[i + 1] = o; });

  for (const track of tracks) {
    const idx   = track.trackIndex;
    const onset = onsetMap[idx];

    if (idx === 0) {
      track.startTime       = 0;
      track.startTimeMethod = 'manual';
      delete track.startTimeConfidence;
    } else if (onset) {
      track.startTime           = Math.round(onset.time * 10) / 10;
      track.startTimeConfidence = onset.confidence;
      track.startTimeMethod     = 'onset-derivative';
    } else {
      delete track.startTime;
      delete track.startTimeConfidence;
      track.startTimeMethod = 'onset-derivative';
    }
  }

  // Compute durationSecs for each track
  for (let i = 0; i < tracks.length; i++) {
    const curr = tracks[i];
    const next = tracks[i + 1];

    if (curr.startTime !== undefined) {
      if (next?.startTime !== undefined) {
        curr.durationSecs = Math.round((next.startTime - curr.startTime) * 10) / 10;
      } else if (i === tracks.length - 1) {
        curr.durationSecs = Math.round((episodeDurationSecs - curr.startTime) * 10) / 10;
      } else {
        curr.durationSecs = null;
      }
    } else {
      curr.durationSecs = null;
    }
  }
}

/**
 * Print a formatted summary table to stdout.
 */
function printTable(tracks, episodeDurationSecs) {
  const header = '  Idx  Track                                                  StartTime  Confidence  Duration';
  console.log(`\nDetected track boundaries for episode #${episodeId} (${trackCount} tracks, ${fmtTime(episodeDurationSecs)} total):\n`);
  console.log(header);
  console.log('  ' + '-'.repeat(header.length - 2));

  const lowConf = [];

  for (const t of tracks) {
    const idxStr   = String(t.trackIndex).padStart(3);
    const name     = `${t.artist} \u2013 ${t.title}`;
    const nameStr  = name.slice(0, 52).padEnd(53);
    const startStr = (t.startTime !== undefined ? fmtTime(t.startTime) : '\u2014').padEnd(10);
    const confStr  = t.trackIndex === 0
      ? '(fixed)   '
      : (t.startTimeConfidence !== undefined
          ? String(t.startTimeConfidence.toFixed(2)).padEnd(10)
          : '?'.padEnd(10));
    const durStr   = (t.durationSecs !== null && t.durationSecs !== undefined)
      ? fmtTime(t.durationSecs)
      : '\u2014';

    console.log(`  ${idxStr}  ${nameStr}${startStr} ${confStr} ${durStr}`);

    if (t.trackIndex > 0 && t.startTimeConfidence !== undefined && t.startTimeConfidence < 0.65) {
      lowConf.push(t.trackIndex);
    }
  }

  console.log('');

  if (lowConf.length > 0) {
    console.log(`  \u26a0  Tracks with confidence < 0.65: #${lowConf.join(', #')}  \u2014 verify by listening`);
  }

  const sumDuration = tracks.reduce((s, t) => s + (t.durationSecs || 0), 0);
  const diff        = Math.abs(sumDuration - episodeDurationSecs);
  const sumOk       = diff < 10;
  console.log(`  Sum check: ${fmtTime(sumDuration)} / ${fmtTime(episodeDurationSecs)} (${sumOk ? '\u2713 OK' : `\u26a0 off by ${diff.toFixed(0)}s`})`);
  console.log('');
}

function fmtTime(secs) {
  if (secs == null) return '\u2014';
  const h  = Math.floor(secs / 3600);
  const m  = Math.floor((secs % 3600) / 60);
  const s  = Math.floor(secs % 60);
  const mm = String(m).padStart(h > 0 ? 2 : 1, '0');
  const ss = String(s).padStart(2, '0');
  return h > 0 ? `${h}:${mm}:${ss}` : `${mm}:${ss}`;
}

File: scripts/find-onsets.py
#!/usr/bin/env python3
"""
find-onsets.py

Two-pass track boundary detection using librosa.

Show structure (confirmed):
  [Opener song] → [Talk] → [Music set] → [Talk] → [Music set] → ... → [Talk] → [Closer song]

Pass 1 – Talk detection:
  Sliding-minimum RMS with adaptive threshold. Talk segments = regions that stay
  well below the music median for at least TALK_MIN_SECS seconds.
  Each gap between talk segments is a music section.

Pass 2 – Within-section onset detection:
  For each middle audio section, detect track boundaries using librosa spectral
  flux with MIN_TRACK_SECS minimum gap. Section starts are always included as
  track starts (they're definitive boundaries after talk).

Track assignment:
  Track 0  → audio section 1 (opener, startTime = 0)
  Track N-1 → audio section -1 (closer, startTime = section start)
  Tracks 1..N-2 → all detected starts in middle sections, assigned in order

Usage:
  python3 scripts/find-onsets.py 1975
  python3 scripts/find-onsets.py 1975 --dry-run
  python3 scripts/find-onsets.py 1975 --verbose
"""

import sys, json, os, warnings
import numpy as np
warnings.filterwarnings('ignore')
import librosa

ROOT      = os.path.join(os.path.dirname(__file__), '..')
DATA_FILE = os.path.join(ROOT, 'public', 'data', 'episodes.json')
AUDIO_DIR = os.path.join(ROOT, 'public', 'audio')

# ── Tuning ─────────────────────────────────────────────────────────────────────
ANALYSIS_SR         = 8000   # downsample for fast loading (Hz)

# Pass 1 – talk segment detection
TALK_WIN_SECS       = 0.5    # short RMS frame (fine resolution)
TALK_HOP_SECS       = 0.25   # 0.25s hop
TALK_SMOOTH_SECS    = 5.0    # sliding-minimum window: need 5s of sustained quiet
TALK_DB_BELOW_MUSIC = 6      # threshold: this many dB below music median
TALK_MIN_SECS       = 30     # discard low-energy regions shorter than this
TALK_MERGE_SECS     = 10     # merge talk segments within this gap

# Pass 2 – within-section boundary detection (RMS valley method)
# Assumes equal track distribution across sections unless TRACKS_PER_SECTION is set.
# The valley method finds the deepest energy dip in each expected inter-track window.
TRACKS_PER_SECTION  = 4      # tracks per middle section (16 middle tracks / 4 sections)
VALLEY_SMOOTH_SECS  = 3.0    # smooth RMS over this window before finding valley
VALLEY_SEARCH_PCT   = 0.40   # search ±this fraction of track window around expected boundary
# ──────────────────────────────────────────────────────────────────────────────


def detect_talk_segments(rms_db, times):
    """
    Find sustained low-energy regions (DJ talking between sets).
    Uses a sliding minimum to suppress brief loud transients within talk.
    Returns: (segments, threshold_db, music_median_db)
    """
    music_half   = rms_db[rms_db > np.percentile(rms_db, 50)]
    music_median = float(np.median(music_half))
    threshold_db = music_median - TALK_DB_BELOW_MUSIC

    hop_secs      = float(times[1] - times[0]) if len(times) > 1 else TALK_HOP_SECS
    smooth_frames = max(1, int(TALK_SMOOTH_SECS / hop_secs))
    smoothed      = np.array([
        np.min(rms_db[max(0, i - smooth_frames // 2): i + smooth_frames // 2 + 1])
        for i in range(len(rms_db))
    ])

    below = smoothed < threshold_db
    segments = []
    start = None
    for i, is_low in enumerate(below):
        if is_low and start is None:
            start = float(times[i])
        elif not is_low and start is not None:
            dur = float(times[i]) - start
            if dur >= TALK_MIN_SECS:
                segments.append((start, float(times[i])))
            start = None
    if start is not None and float(times[-1]) - start >= TALK_MIN_SECS:
        segments.append((start, float(times[-1])))

    # Merge nearby
    merged = []
    for seg in segments:
        if merged and seg[0] - merged[-1][1] < TALK_MERGE_SECS:
            merged[-1] = (merged[-1][0], seg[1])
        else:
            merged.append(list(seg))

    return [(s, e) for s, e in merged], threshold_db, music_median


def talk_to_sections(talk_segs, duration_secs):
    """
    Convert talk-segment list into music section (start, end) pairs.
    Each section is the audio between talk segments.
    """
    boundaries = [0.0]
    for ts, te in talk_segs:
        boundaries.append(ts)
        boundaries.append(te)
    boundaries.append(duration_secs)

    sections = []
    for i in range(0, len(boundaries) - 1, 2):
        sections.append((boundaries[i], boundaries[i + 1]))
    return sections


def detect_section_onsets(y, sr, sec_start, sec_end, n_tracks, verbose=False):
    """
    Find n_tracks-1 track boundaries within [sec_start, sec_end] using RMS
    valley detection. Assumes roughly equal-length tracks; searches for the
    deepest energy dip in a window centred on each expected boundary.

    This is more reliable than spectral onset detection for back-to-back punk
    tracks: any brief gap between songs (even 0.2s) shows up as a valley.
    """
    s0  = int(sec_start * sr)
    s1  = int(sec_end   * sr)
    seg = y[s0:s1]
    dur = sec_end - sec_start

    if n_tracks <= 1 or dur < 30:
        return []

    hop          = 256                         # ~32ms at 8kHz
    frame_length = 512
    rms   = librosa.feature.rms(y=seg, frame_length=frame_length, hop_length=hop)[0]
    times = librosa.frames_to_time(np.arange(len(rms)), sr=sr, hop_length=hop)

    # Smooth to suppress within-song transients (VALLEY_SMOOTH_SECS window)
    smooth_frames = max(1, int(VALLEY_SMOOTH_SECS * sr / hop))
    smoothed      = np.convolve(rms, np.ones(smooth_frames) / smooth_frames, mode='same')

    n_boundaries = n_tracks - 1
    track_window = dur / n_tracks   # expected duration per track

    boundaries = []
    for b in range(n_boundaries):
        expected   = (b + 1) * track_window
        search_lo  = max(10, expected - VALLEY_SEARCH_PCT * track_window)
        search_hi  = min(dur - 10, expected + VALLEY_SEARCH_PCT * track_window)

        mask = (times >= search_lo) & (times < search_hi)
        if not mask.any():
            if verbose:
                print(f'    boundary {b+1}: no frames in search window [{fmt_time(sec_start+search_lo)}–{fmt_time(sec_start+search_hi)}]')
            continue

        min_idx   = int(np.argmin(smoothed[mask]))
        abs_idx   = int(np.where(mask)[0][min_idx])
        found_t   = float(times[abs_idx])
        valley_db = float(librosa.amplitude_to_db(np.array([smoothed[abs_idx]]), ref=np.max(smoothed))[0])

        if verbose:
            print(f'    boundary {b+1}: expected {fmt_time(sec_start+expected)}, '
                  f'valley at {fmt_time(sec_start+found_t)} ({valley_db:.1f} dB)')

        boundaries.append(sec_start + found_t)

    return boundaries


def fmt_time(secs):
    if secs is None:
        return '—'
    secs = int(round(secs))
    h = secs // 3600
    m = (secs % 3600) // 60
    s = secs % 60
    return f'{h}:{m:02d}:{s:02d}' if h else f'{m}:{s:02d}'


def main():
    args    = sys.argv[1:]
    ep_arg  = next((a for a in args if a.isdigit()), None)
    dry_run = '--dry-run' in args
    verbose = '--verbose' in args

    if not ep_arg:
        print('Usage: python3 scripts/find-onsets.py <episodeId> [--dry-run] [--verbose]',
              file=sys.stderr)
        sys.exit(1)

    ep_id = int(ep_arg)
    with open(DATA_FILE) as f:
        data = json.load(f)

    episode = next((e for e in data['episodes'] if e['id'] == ep_id), None)
    if not episode:
        print(f'Episode {ep_id} not found', file=sys.stderr)
        sys.exit(1)

    mp3 = os.path.join(AUDIO_DIR, f'mrr-radio-{ep_id}.mp3')
    if not os.path.exists(mp3):
        print(f'MP3 not found: {mp3}', file=sys.stderr)
        sys.exit(1)

    tracks      = sorted(episode.get('tracks', []), key=lambda t: t['trackIndex'])
    duration    = episode.get('durationSecs', 0)
    track_count = len(tracks)

    print(f'\nEpisode #{ep_id}: {track_count} tracks, {fmt_time(duration)} total')
    print(f'Loading audio at {ANALYSIS_SR} Hz...')
    y, sr = librosa.load(mp3, sr=ANALYSIS_SR, mono=True)
    print(f'Loaded {len(y)/sr:.0f}s\n')

    # ── Pass 1: Talk detection ─────────────────────────────────────────────────
    print('── Pass 1: Talk detection ───────────────────────────────────────────────')
    win      = int(TALK_WIN_SECS * sr)
    hop      = int(TALK_HOP_SECS * sr)
    rms      = librosa.feature.rms(y=y, frame_length=win, hop_length=hop)[0]
    rms_db   = librosa.amplitude_to_db(rms, ref=np.max)
    rms_t    = librosa.frames_to_time(np.arange(len(rms)), sr=sr, hop_length=hop)

    talk_segs, threshold_db, music_median = detect_talk_segments(rms_db, rms_t)
    audio_sections = talk_to_sections(talk_segs, duration)

    print(f'  Music baseline: {music_median:.1f} dB  |  talk threshold: {threshold_db:.1f} dB')
    print(f'  {len(talk_segs)} talk segment(s):')
    for ts, te in talk_segs:
        print(f'    {fmt_time(ts)} – {fmt_time(te)}  ({te - ts:.0f}s)')
    print(f'  → {len(audio_sections)} music section(s):')
    for i, (ss, se) in enumerate(audio_sections):
        print(f'    [{i+1}] {fmt_time(ss)} – {fmt_time(se)}  ({se - ss:.0f}s)')

    # Validate: expect at least opener + closer + 1 middle section
    if len(audio_sections) < 3:
        print('\n⚠  Too few sections detected. Adjust TALK_DB_BELOW_MUSIC and re-run.')
        sys.exit(1)

    opener_section = audio_sections[0]
    closer_section = audio_sections[-1]
    middle_sections = audio_sections[1:-1]

    # ── Pass 2: Within-section onset detection ────────────────────────────────
    print('\n── Pass 2: Within-section onset detection ───────────────────────────────')

    # Collect all middle track starts: each section start + within-section valleys
    all_middle_starts = []
    for i, (sec_start, sec_end) in enumerate(middle_sections):
        all_middle_starts.append(sec_start)  # section start = definite track start
        valleys = detect_section_onsets(y, sr, sec_start, sec_end,
                                        n_tracks=TRACKS_PER_SECTION, verbose=verbose)
        dur_str = fmt_time(sec_end - sec_start)
        print(f'  Section {i+2}: {fmt_time(sec_start)}–{fmt_time(sec_end)} ({dur_str})'
              f'  →  {len(valleys)} valley(s): '
              + (', '.join(fmt_time(v) for v in valleys) or 'none'))
        all_middle_starts.extend(valleys)

    all_middle_starts.sort()

    # ── Assign start times ────────────────────────────────────────────────────
    start_times   = {}
    start_methods = {}

    # Track 0: opener
    start_times[0]   = 0.0
    start_methods[0] = 'manual'

    # Track N-1: closer
    closer_idx = track_count - 1
    start_times[closer_idx]   = closer_section[0]
    start_methods[closer_idx] = 'talk-boundary'

    # Middle tracks: assign detected starts in order
    middle_tracks = [t for t in tracks if 0 < t['trackIndex'] < closer_idx]
    n_need  = len(middle_tracks)
    n_found = len(all_middle_starts)

    if n_found != n_need:
        print(f'\n  ⚠  Found {n_found} start times for {n_need} middle tracks.')
        if n_found < n_need:
            print(f'     Lower MIN_TRACK_SECS (currently {MIN_TRACK_SECS}s) or ONSET_DELTA and re-run.')
        else:
            print(f'     Raise MIN_TRACK_SECS (currently {MIN_TRACK_SECS}s) and re-run.')

    for i, track in enumerate(middle_tracks):
        if i < n_found:
            start_times[track['trackIndex']]   = all_middle_starts[i]
            start_methods[track['trackIndex']] = 'talk-boundary' if all_middle_starts[i] in [s for s, _ in middle_sections] else 'onset-librosa'
        else:
            start_times[track['trackIndex']]   = None
            start_methods[track['trackIndex']] = 'onset-librosa'

    # ── Compute durations and write fields back ───────────────────────────────
    for i, track in enumerate(tracks):
        idx   = track['trackIndex']
        start = start_times.get(idx)
        meth  = start_methods.get(idx, '?')

        next_start = None
        for j in range(idx + 1, track_count):
            ns = start_times.get(j)
            if ns is not None:
                next_start = ns
                break
        dur = None
        if start is not None:
            dur = (next_start - start) if next_start is not None else (duration - start)

        if idx == 0:
            conf = None
        elif meth == 'talk-boundary':
            conf = 0.90
        else:
            conf = 0.65
        if dur is not None and dur < 60 and conf:
            conf = min(conf, 0.40)

        if start is not None:
            track['startTime']       = round(start, 1)
            track['startTimeMethod'] = meth
            if conf is not None:
                track['startTimeConfidence'] = conf
            elif 'startTimeConfidence' in track:
                del track['startTimeConfidence']
        else:
            track.pop('startTime', None)
            track.pop('startTimeConfidence', None)
            track['startTimeMethod'] = meth
        track['durationSecs'] = round(dur, 1) if dur is not None else None

    # ── Results table (tracks interleaved with DJ segments) ───────────────────
    print('\n── Results ──────────────────────────────────────────────────────────────')
    print(f'  {"":52}  {"Start":>7}  {"End":>7}  {"Dur":>6}  Note')
    print('  ' + '─' * 85)

    # Build a flat timeline: tracks + DJ segments, sorted by start time
    timeline = []
    for track in tracks:
        s = track.get('startTime')
        d = track.get('durationSecs')
        e = (s + d) if (s is not None and d is not None) else None
        meth = track.get('startTimeMethod', '?')
        conf = track.get('startTimeConfidence')
        flag = ' ⚠' if (d is not None and d < 60) else ''
        conf_tag = f'conf={conf:.2f}' if conf is not None else 'fixed'
        timeline.append(('track', s, e, track, f'{conf_tag}{flag}'))
    for ts, te in talk_segs:
        timeline.append(('dj', ts, te, None, f'{te-ts:.0f}s'))

    timeline.sort(key=lambda x: (x[1] is None, x[1] or 0))

    for kind, start, end, track, note in timeline:
        start_str = fmt_time(start)
        end_str   = fmt_time(end)
        dur_secs  = (end - start) if (start is not None and end is not None) else None
        dur_str   = fmt_time(dur_secs)
        if kind == 'dj':
            label = f'  {"── DJ ──":<52}'
            print(f'{label}  {start_str:>7}  {end_str:>7}  {dur_str:>6}  {note}')
        else:
            idx   = track['trackIndex']
            name  = f'#{idx:<2} {track["artist"]} – {track["title"]}'
            print(f'  {name[:52]:<52}  {start_str:>7}  {end_str:>7}  {dur_str:>6}  {note}')

    sum_dur = sum(t.get('durationSecs') or 0 for t in tracks)
    diff    = abs(sum_dur - duration)
    ok      = '✓ OK' if diff < 10 else f'⚠ off by {diff:.0f}s'
    print(f'\n  Music sum: {fmt_time(sum_dur)} / {fmt_time(duration)} episode ({ok})')

    short = [t for t in tracks if t.get('durationSecs') and t['durationSecs'] < 60]
    if short:
        print(f'  ⚠  Under 1 minute — verify by ear:')
        for t in short:
            print(f'     #{t["trackIndex"]} {t["artist"]} ({fmt_time(t.get("durationSecs"))})')

    print()
    if not dry_run:
        with open(DATA_FILE, 'w') as f:
            json.dump(data, f, indent=2)
        print('Written to episodes.json.')
    else:
        print('--dry-run: episodes.json not modified.')
    print()


if __name__ == '__main__':
    main()
